lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb

```diff
@@ -1048,22 +1042,31 @@ def cmd_upload_tabs(str, words)
   # sometimes it wouldn't execute successfully especailly on bad network.
   #
   def tab_complete_cfilenames(str, words)
-    if client.commands.include?(COMMAND_ID_STDAPI_FS_LS)
-      return client.fs.dir.match(str) rescue nil
-    end
-
-    []
+    tab_complete_path(str, words, false)
   end
 
   #
   # Provide a generic tab completion for client directory names.
   #
   def tab_complete_cdirectory(str, words)
+    tab_complete_path(str, words, true)
+  end
+
+  def tab_complete_path(str, words, dir_only)
     if client.commands.include?(COMMAND_ID_STDAPI_FS_LS)
-      return client.fs.dir.match(str, true) rescue nil
+      results = client.fs.dir.match(str, dir_only) rescue []
+      if results.length == 1 && results[0] != str && results[0].end_with?(client.fs.file.separator)
+        # If Readline receives a single value from this function, it will assume we're done with the tab
+        # completing, and add an extra space at the end.
+        # This is annoying if we're recursively tab-traversing our way through subdirectories -
+        # we may want to continue traversing, but MSF will add a space, requiring us to back up to continue
+        # tab-completing our way through successive subdirectories.
+        ::Readline.completion_append_character = nil
+      end
+      results
+    else
+      []
     end
-
-    []
   end
```

Comment:
Could remove the else condition since its not needed and if the `if` statement isn't hit it will fall through to the `else` one. I think this would be more Ruby like.